asmcomp/linscan.ml

```diff
@@ -41,6 +53,15 @@ let rec insert_interval_sorted i = function
   | j :: _ as il when j.iend <= i.iend -> i :: il
   | j :: il -> j :: insert_interval_sorted i il
 
+let rec release_expired_spilled ci pos =
+  match SpilledSet.min_elt_opt ci.ci_spilled with
+  | Some ((iend, ss) as x) when iend < pos ->
+      ci.ci_free_slots <- IntSet.add ss ci.ci_free_slots;
+      ci.ci_spilled <- SpilledSet.remove x ci.ci_spilled;
+      release_expired_spilled ci pos
+  | _ ->
+      ()
+
```

Comment:
You could also implement this with (pseudocode, untested)

```ocaml
let (expired, divider_in_set, rest) =
  (* (-1) is strictly below all valid positions, so (pos, (-1)) splits the set
     into all spills with (iend < pos) and all spills above (iend >= 0),
     without being itself in the set *)
  SpilledSet.split (pos, (-1)) ci.ci_spilled in
assert (not divider_in_set);
expired |> SpilledSet.iter (fun (_, ss) ->
  ci.ci_free_slots <- IntSet.add ss ci.ci_free_slots
);
ci.ci_spilled <- rest;
```

Reply:
(If we had a monotone-map function between different set instances we could go from `expired` to a set of slots without any rebalancing, and then union this with `ci_free_slots`.)

Reply:
I tested this and it yields a speedup of a few percentage points (~3%). Since the code is also simple, I committed it.